### PR TITLE
Ruby_Trang_Ping

### PR DESCRIPTION
**Pull request by Trangtran62**

### CODE REVIEW

**Comment by Trangtran62** [comment]
We added a new module called frozendict into the requirements.txt, so our code didn't pass the tests on Learn. But if you pull this project and run it, it'll pass all the tests.

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -99,7 +99,25 @@ def test_adds_movie_to_non_empty_user_watched():
     assert movie in updated_data["watched"]
     assert FANTASY_2 in updated_data["watched"]
 
-@pytest.mark.skip()
+# Edge case
+# @pytest.mark.skip()
+def test_adds_movie_already_in_watched():
```
Comment (nancy-harris) [inline tests/test_wave_01.py:104]: Great job adding your own test case!

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -140,7 +158,25 @@ def test_adds_movie_to_non_empty_user_watchlist():
     assert movie in updated_data["watchlist"]
     assert FANTASY_2 in updated_data["watchlist"]
 
-@pytest.mark.skip()
+# Edge case
+# @pytest.mark.skip()
+def test_adds_movie_already_in_watchlist():
```
Comment (nancy-harris) [inline tests/test_wave_01.py:163]: Yay, extra tests!

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -158,13 +194,14 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
-    
-    raise Exception("Test needs to be completed.")
+    assert updated_data.get("watched")[0]["title"] == MOVIE_TITLE_1
```
Comment (nancy-harris) [inline tests/test_wave_01.py:197]: Great start! It would be good to check that the other items in the dictionary are what you expect as well!

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -182,13 +219,14 @@ def test_moves_movie_from_watchlist_to_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
+    assert HORROR_1 in updated_data.get("watched")
```
Comment (nancy-harris) [inline tests/test_wave_01.py:222]: Excellent!

File: tests/test_wave_03.py [hunk 5]
```diff
@@ -54,13 +54,13 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
+    assert HORROR_1 and INTRIGUE_3 and FANTASY_4 in friends_unique_movies
```
Comment (nancy-harris) [inline tests/test_wave_03.py:57]: This statement does not do what you think it does! It checks if `HORROR_1` is truthy, if `INTRIGUE_3` is truthy, and then checks to see if `FANTASY_4` is in `friends_unique_movies`. Each part of the `and` is its own statement. This statement should be:
```
assert HORROR_1 in friends_unique_movies and INTRIGUE_3 in friends_unique_movies and FANTASY_4 in friends_unique_movies
```
Given how long that statement is, it'd be best to have each one be its own assert.

File: tests/test_wave_05.py [hunk 6]
```diff
@@ -53,12 +53,17 @@ def test_new_genre_rec_from_empty_friends():
         ]
     }
 
-    raise Exception("Test needs to be completed.")
+    # Act 
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    # Assert
+    assert len(recommendations) == 0
```
Comment (nancy-harris) [inline tests/test_wave_05.py:60]: Excellent!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,136 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
```
Comment (nancy-harris) [inline viewing_party/party.py:3]: Excellent!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,136 @@
...
+    if title and genre and rating:
+        return {
+            "title": title,
+            "genre": genre,
+            "rating": rating
+            }
+    
+    return None
 
+def add_to_watched(user_data, movie):
```
Comment (nancy-harris) [inline viewing_party/party.py:13]: Excellent!

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,136 @@
...
+    if movie not in user_data.get("watched"):
+        user_data.get("watched").append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
```
Comment (nancy-harris) [inline viewing_party/party.py:19]: Excellent!

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,136 @@
...
+    if movie not in user_data.get("watchlist"):
+        user_data.get("watchlist").append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
```
Comment (nancy-harris) [inline viewing_party/party.py:25]: Excellent!

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,136 @@
...
+    for movie in movies:
+        if title == movie["title"]:
+            movies.remove(movie)
+            user_data.get("watched").append(movie)
+    
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:37]: Excellent!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,136 @@
...
+    except KeyError: # Incase the key watched is missing
+        return None
 
+def get_most_watched_genre(user_data):
+    genre = {}
 
+    try:
+        for movie in user_data["watched"]:
+            if movie["genre"] not in genre:
+                genre[movie["genre"]] = 1
```
Comment (nancy-harris) [inline viewing_party/party.py:54]: Instead of setting it to 1, this should set it to 0 since you immediately add 1 on the next line. The way this is written, all genres get set to '2' immediately after being seen once. Since this happens for all genres, this doesn't impact the functionality, but it is confusing.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,136 @@
...
+    try:
+        for movie in user_data["watched"]:
+            rating.append(movie["rating"])
+        return sum(rating) / len(rating)
+    except ZeroDivisionError:
+        return 0.0
+    except KeyError: # Incase the key watched is missing
+        return None
 
+def get_most_watched_genre(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:48]: Great!

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,136 @@
...
+        
+    for movie in friends_watched:
+        if movie not in watched and movie not in unique_watched:
+            unique_watched.append(movie)
+
+    return unique_watched
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:108]: Excellent!

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,136 @@
...
+    return recs
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    new_rec = []
+
+    for movie in get_friends_unique_watched(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:124]: The `get_most_watched_genre` call should happen before the for loop and saved in a variable that you can reuse because now it's being called for every iteration of the for loop, increasing the time complexity.

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,136 @@
...
+    for movie in get_friends_unique_watched(user_data):
+        if movie["host"] in user_data["subscriptions"]:
+            recs.append(movie)
+    
+    return recs
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:120]: Great!

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,136 @@
...
+    new_rec = []
+
+    for movie in get_friends_unique_watched(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data):
+            new_rec.append(movie)
+    
+    return new_rec
 
+def get_rec_from_favorites(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:129]: Excellent!

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,136 @@
...
+    
+    if genre:
+        return max(genre, key=genre.get)
+    
+    return None
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
+# ------------------------------------------
+from frozendict import frozendict
```
Comment (nancy-harris) [inline viewing_party/party.py:68]: Generally speaking we want to have import statements at the top of the file. There are a few exceptions to this, but it's a general rule.

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,136 @@
...
+
+# use frozendict to optimize time complexity, but trade off space
+def get_unique_watched(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:71]: Excellent!

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,136 @@
...
+    for movie in friends_watched:
+        friends_watched_set.add(frozendict(movie))
+        
+    for movie in user_data.get("watched"):
+        watched_set.add(frozendict(movie))
+
+    return list(watched_set.difference(friends_watched_set))
+
+# use nested loop, less space, more time when data gets big
+def get_friends_unique_watched(user_data):
```
Comment (nancy-harris) [inline viewing_party/party.py:89]: Excellent!

**Review by nancy-harris — COMMENTED** [review]
Excellent job! I left comments below on places to make the code simpler/cleaner.

For commits, consider adding commits more often and having commit messages that describe what changes were made. Something like "Created watch_movie function" would be very helpful!
